File: rust/parity/specs/runtime_anomaly_sentry.py
"""Parity spec for PiRuntimeAnomalySentry.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiRuntimeAnomalySentry"

_mod = load_py_agent("pi_runtime_anomaly_sentry.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiRuntimeAnomalySentry()
    out = agent.audit_runtime(_mod.RuntimeInput(**data))
    return out.model_dump()


_CLEAN = "cpu: 12%\nmemory: 40%\nerror_rate: 0%\nthreads: nominal"
_CPU = "cpu_spike detected on worker-3\nmemory ok"
_OOM = "container restarted: OOM_KILLED\nnode pressure high"
_ERROR = "error_rate: 45%\nlatency rising"
_5XX = "5xx_errors: high across edge nodes"
_SHELL = "unauthorized outbound to 8.8.8.8 via sh: /bin/sh -c"
_CMD = "suspicious connection to evil.example.com spawning cmd.exe"
# uppercase to exercise .lower() normalization
_UPPER = "Detected SUSPICIOUS CONNECTION TO 10.0.0.1 launching CMD.EXE"
# multiple anomalies -> risk_score is the max across all matches
_MULTI = "cpu_spike\nerror_rate: 45%\nunauthorized outbound traffic"

SAMPLES = [
    {"input": {"metrics_content": _CLEAN}},
    {"input": {"metrics_content": _CPU}},
    {"input": {"metrics_content": _OOM}},
    {"input": {"metrics_content": _ERROR}},
    {"input": {"metrics_content": _5XX}},
    {"input": {"metrics_content": _SHELL}},
    {"input": {"metrics_content": _CMD}},
    {"input": {"metrics_content": _UPPER}},
    {"input": {"metrics_content": _MULTI}},
    {"input": {"metrics_content": ""}},
    # non-strict env -> WARN_ANOMALIES path, is_secure stays True
    {"input": {"metrics_content": _CPU},
     "env": {"PI_RUNTIME_STRICT_MODE": "false"}},
    {"input": {"metrics_content": _CPU},
     "env": {"PI_RUNTIME_STRICT_MODE": "true"}},
]

File: rust/crates/pi-agents/src/agents/runtime_anomaly_sentry.rs
//! Port of `pi_micro_agents/pi_runtime_anomaly_sentry.py`.
//!
//! Flags runtime metric drift, unauthorized execution binaries, or suspicious
//! outbound connections in production. Behaviour is a line-for-line mirror of
//! the Python original.

use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub metrics_content: String,
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub anomalies_detected: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

/// Mirrors `is_strict_mode()`: strict unless the env var is set to a value that
/// is not (case-insensitively) "true".
fn is_strict_mode() -> bool {
    match std::env::var("PI_RUNTIME_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

pub fn audit_runtime(input: &Input) -> Output {
    let content = input.metrics_content.to_lowercase();
    let mut anomalies: Vec<String> = Vec::new();
    let mut risk_score: f64 = 0.0;

    // CPU / memory spikes
    if content.contains("cpu_spike") || content.contains("cpu: 99%") || content.contains("oom_killed")
    {
        anomalies.push(
            "Resource Exhaustion: Runtime logs indicate critical CPU threshold breaches or container OOM terminations."
                .to_string(),
        );
        risk_score = risk_score.max(70.0);
    }

    // High error rates
    if content.contains("error_rate: 45%") || content.contains("5xx_errors: high") {
        anomalies.push(
            "Uncontrolled Fault Rate: High density of 5xx HTTP exceptions suggests dynamic system instability."
                .to_string(),
        );
        risk_score = risk_score.max(80.0);
    }

    // Unauthorized outbound network connection attempts
    if content.contains("unauthorized outbound")
        || content.contains("suspicious connection to")
        || content.contains("sh: ")
        || content.contains("cmd.exe")
    {
        anomalies.push(
            "Suspicious Shell Execution: Runtime detected unauthorized bash/cmd spawn queries or unexpected ports."
                .to_string(),
        );
        risk_score = risk_score.max(95.0);
    }

    let mut is_sec = true;
    if risk_score > 30.0 && is_strict_mode() {
        is_sec = false;
    }

    let mut status = if is_sec {
        "PASSED".to_string()
    } else {
        "ANOMALIES_DETECTED".to_string()
    };
    if risk_score > 0.0 && is_sec {
        status = "WARN_ANOMALIES".to_string();
    }

    Output {
        is_secure: is_sec,
        anomalies_detected: anomalies,
        risk_score,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = audit_runtime(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(content: &str) -> Output {
        audit_runtime(&Input {
            metrics_content: content.into(),
        })
    }

    #[test]
    fn clean_metrics_pass() {
        let o = run("cpu: 12%\nmemory: 40%\nerror_rate: 0%");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.anomalies_detected.is_empty());
    }

    #[test]
    fn shell_execution_flagged() {
        // case-insensitive: "CMD.EXE" matches "cmd.exe" after lowercasing
        let o = run("Detected SUSPICIOUS CONNECTION TO 10.0.0.1 spawning CMD.EXE");
        assert!(!o.is_secure);
        assert_eq!(o.status, "ANOMALIES_DETECTED");
        assert_eq!(o.risk_score, 95.0);
        assert_eq!(o.anomalies_detected.len(), 1);
    }

    #[test]
    fn cpu_spike_takes_max_risk() {
        // multiple anomalies; risk_score is the max across matches
        let o = run("cpu_spike detected\nerror_rate: 45%");
        assert!(!o.is_secure);
        assert_eq!(o.risk_score, 80.0);
        assert_eq!(o.anomalies_detected.len(), 2);
        assert_eq!(o.status, "ANOMALIES_DETECTED");
    }
}
